src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -759,7 +759,9 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
 
             BoundExpression expression = BindValue(initializer, diagnostics, valueKind);
 
-            if (expression is BoundStackAllocArrayCreation boundStackAlloc)
+            if (expression is BoundStackAllocArrayCreation boundStackAlloc &&
```

Comment:
>if [](start = 12, length = 2)

I didn't understand the purpose of this change.

Reply:
Now that we're relaxing the locations where a stackalloc expression can appear, we need to give it a natural type of a pointer type only in the syntactic context where that is supposed to work. The old code did that even in erroneous locations because it didn't matter. The first test ensures we are in a local variable initializer. The second test ensures that it is not enclosed in some kind of no-op expression like a parenthesized expression, checked, etc, which removes it from the context where it should be treated as a pointer.

---
In reply to: [207692207](https://github.com/dotnet/roslyn/pull/28969#discussion_r207692207) [](ancestors = 207692207)